src/minichain/storage.py

```
@@ -0,0 +1,280 @@
+    def save_state(self, state: State, *, connection: sqlite3.Connection | None = None) -> None:
+        """Persist all accounts as the current canonical state snapshot."""
+        if connection is None:
+            with self._connection:
+                self.save_state(state, connection=self._connection)
+            return
+
+        conn = connection
+
+        for address, account in state.accounts.items():
+            if not _is_lower_hex(address, 40):
+                raise StorageError(f"Invalid account address: {address}")
```

Ruff (0.15.2)

[warning] Avoid specifying long messages outside the exception class

(TRY003)